packages/router/src/grpc/view-protocol-server/delegations-by-address-index.ts

```diff
@@ -0,0 +1,116 @@
+import { ValidatorInfo } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/component/stake/v1/stake_pb';
+import {
+  AssetMetadataByIdRequest,
+  BalancesRequest,
+  BalancesResponse,
+  DelegationsByAddressIndexRequest_Filter,
+  DelegationsByAddressIndexResponse,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1/view_pb';
+import { stakingClientCtx } from '../../ctx';
+import { balances } from './balances';
+import {
+  Metadata,
+  ValueView,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1/asset_pb';
+import { assetMetadataById } from './asset-metadata-by-id';
+
+const isDelegationBalance = (balance: BalancesResponse, identityKey: IdentityKey) => {
+  const match = assetPatterns.delegationToken.exec(getDisplayDenomFromView(balance.balanceView));
+  if (!match) return false;
+
+  const matchGroups = match.groups as unknown as DelegationCaptureGroups;
+
+  return bech32IdentityKey(identityKey) === matchGroups.bech32IdentityKey;
+};
+
+const getDelegationTokenBaseDenom = (validatorInfo: ValidatorInfo) =>
+  `udelegation_${bech32IdentityKey(getIdentityKeyFromValidatorInfo(validatorInfo))}`;
+
+const addressHasDelegationTokens = (
+  delegation?: PartialMessage<BalancesResponse>,
+): delegation is PartialMessage<BalancesResponse> & { balanceView: ValueView } =>
+  delegation?.balanceView instanceof ValueView;
+
+export const delegationsByAddressIndex: Impl['delegationsByAddressIndex'] = async function* (
+  req,
+  ctx,
+) {
+  const { addressIndex } = req;
+  if (!addressIndex) {
+    throw new Error('Missing `addressIndex` in `DelegationsByAddressIndex` request');
+  }
+
+  const stakingClient = ctx.values.get(stakingClientCtx);
+  if (!stakingClient) throw new Error('Staking context not found');
+
+  const assetBalances = await Array.fromAsync(
+    balances(new BalancesRequest({ accountFilter: addressIndex }), ctx),
+  );
+
+  for await (const validatorInfoResponse of stakingClient.validatorInfo({
+    showInactive: req.filter === DelegationsByAddressIndexRequest_Filter.ALL,
+  })) {
+    const validatorInfo = getValidatorInfo(validatorInfoResponse);
+    const extendedMetadata = new Any({
+      typeUrl: ValidatorInfo.typeName,
+      value: validatorInfo.toBinary(),
+    });
+
+    const identityKey = getValidatorInfo.pipe(getIdentityKeyFromValidatorInfo)(
+      validatorInfoResponse,
+    );
+    const delegation = assetBalances.find(balance =>
+      isDelegationBalance(new BalancesResponse(balance), identityKey),
+    );
+
+    if (addressHasDelegationTokens(delegation)) {
+      const withValidatorInfo = delegation.balanceView.clone();
+
+      if (withValidatorInfo.valueView.case !== 'knownAssetId')
+        throw new Error(`Unexpected ValueView case: ${withValidatorInfo.valueView.case}`);
+
+      withValidatorInfo.valueView.value.extendedMetadata = extendedMetadata;
+
+      yield new DelegationsByAddressIndexResponse({ valueView: withValidatorInfo });
```

Comment:
yielding a `DelegationsByAddressIndexResponse` instead of a `ValueView`, like we were previously doing